### ISSUE TITLE
Seg Fault on cluster shutdown

### ISSUE THREAD

**Issue by richardliaw** [ray-project/ray#6025]
reproduce easily with 
`pytest ray/python/ray/tune/tests/test_cluster.py -k test_cluster_down_simple`

```
F1026 07:30:07.625059 26700 core_worker.cc:216]  Check failed: _s.ok() Bad status: IOError: [RayletClient] Connection closed unexpectedly.
*** Check failure stack trace: ***
    @     0x7fc721e409dd  google::LogMessage::Fail()
    @     0x7fc721e42723  google::LogMessage::SendToLog()
    @     0x7fc721e4053a  google::LogMessage::Flush()
    @     0x7fc721e40771  google::LogMessage::~LogMessage()
    @     0x7fc721a80f48  ray::RayLog::~RayLog()
    @     0x7fc72191a825  ray::CoreWorker::ReportActiveObjectIDs()
    @     0x7fc72190c583  boost::asio::detail::wait_handler<>::do_complete()
    @     0x7fc721911fb4  boost::asio::detail::scheduler::run()
    @     0x7fc721912313  ray::CoreWorker::RunIOService()
    @     0x7fc7211abc5c  execute_native_thread_routine_compat
    @     0x7fc72be6f6ba  start_thread
    @     0x7fc72bba541d  clone
Aborted (core dumped)
```

Here is the core-dump:

```
ubuntu@ip-172-31-43-246:~/ray/python/ray/tune/tests$ cat python-driver-01000000ffffffffffffffffffffffffffffffff.ip-172-31-43-246.ubuntu.log.WARNING.python-driver-01000000ffffffffffffffffffffffffffffffff20191026-075835.27118
Log file created at: 2019/10/26 07:58:35
Running on machine: ip-172-31-43-246
Log line format: [IWEF]mmdd hh:mm:ss.uuuuuu threadid file:line] msg
W1026 07:58:35.568866 27118 grpc_server.cc:16] No service is found when start grpc server driver
W1026 07:58:35.573628 27227 core_worker.cc:200] Tried to erase non-existent object ID6a9388b63062f61c1637c7929d39a0c3ba2a773c
ubuntu@ip-172-31-43-246:~/ray/python/ray/tune/tests$
```

**Comment by richardliaw** [comment]
Looks something like:
```
def test_cluster_down_simple(start_connected_cluster, tmpdir):
    """Tests that TrialRunner save/restore works on cluster shutdown."""
    cluster = start_connected_cluster
    cluster.add_node(num_cpus=1)
    cluster.wait_for_nodes()

    dirpath = str(tmpdir)
    runner = TrialRunner(local_checkpoint_dir=dirpath, checkpoint_period=0)
    kwargs = {
        "stopping_criterion": {
            "training_iteration": 2
        },
        "checkpoint_freq": 1,
        "max_failures": 1
    }
    trials = [Trial("__fake", **kwargs), Trial("__fake", **kwargs)]
    for t in trials:
        runner.add_trial(t)

    runner.step()  # start
    runner.step()  # start2
    runner.step()  # step
    assert all(t.status == Trial.RUNNING for t in runner.get_trials())
    runner.checkpoint()

    cluster.shutdown()
    ray.shutdown()

    cluster = _start_new_cluster()
    runner = TrialRunner(resume="LOCAL", local_checkpoint_dir=dirpath)
    runner.step()  # start
    runner.step()  # start2

    for i in range(3):
        runner.step()

    with pytest.raises(TuneError):
        runner.step()

    assert all(t.status == Trial.TERMINATED for t in runner.get_trials())
    cluster.shutdown()
```

Fails on the last line. If a `sleep(2)` is added right before, it does not fail.

@edoakes

**Comment by edoakes** [comment]
This should be fixed by #6014. Will merge that in the next day or two.
